File: backend/src/routes/hotels.js
const express = require("express");
const router = express.Router();
const { hotels, bookings, reviews: allReviews } = require("../data/hotels");

// GET /api/hotels — search & filter
router.get("/", (req, res) => {
  const {
    city, amenities, minPrice, maxPrice, stars, type, sort, q,
    checkIn, checkOut, country, page: pageStr, limit: limitStr,
  } = req.query;

  let results = [...hotels];

  // Free-text search
  if (q) {
    const lq = q.toLowerCase();
    results = results.filter(
      (h) =>
        h.name.toLowerCase().includes(lq) ||
        h.location.toLowerCase().includes(lq) ||
        h.description.toLowerCase().includes(lq) ||
        h.amenities.some((a) => a.toLowerCase().includes(lq))
    );
  }

  // City filter
  if (city) {
    results = results.filter((h) =>
      h.city.toLowerCase().includes(city.toLowerCase())
    );
  }

  // Country filter
  if (country) {
    const lc = country.toLowerCase();
    results = results.filter((h) => h.country.toLowerCase().includes(lc));
  }

  // Amenity filter
  if (amenities) {
    const requested = amenities.split(",").map((a) => a.trim().toLowerCase());
    results = results.filter((h) =>
      requested.every((ra) =>
        h.amenities.some((ha) => ha.toLowerCase().includes(ra))
      )
    );
  }

  // Price range
  if (minPrice) results = results.filter((h) => h.pricePerNight >= Number(minPrice));
  if (maxPrice) results = results.filter((h) => h.pricePerNight <= Number(maxPrice));

  // Stars filter
  if (stars) {
    results = results.filter((h) => h.stars >= Number(stars));
  }

  // Property type — support comma-separated
  if (type) {
    const types = type.split(",").map((t) => t.trim().toLowerCase());
    results = results.filter((h) => types.includes(h.type.toLowerCase()));
  }

  // Availability check — filter out hotels with overlapping bookings
  if (checkIn && checkOut) {
    const inDate = new Date(checkIn);
    const outDate = new Date(checkOut);
    if (!isNaN(inDate) && !isNaN(outDate) && outDate > inDate) {
      results = results.filter((h) => {
        const overlapping = bookings.filter(
          (b) =>
            b.hotelId === h.id &&
            b.status !== "cancelled" &&
            new Date(b.checkIn) < outDate &&
            new Date(b.checkOut) > inDate
        );
        return overlapping.length === 0;
      });
    }
  }

  // Sorting
  switch (sort) {
    case "price_asc":
      results.sort((a, b) => a.pricePerNight - b.pricePerNight);
      break;
    case "price_desc":
      results.sort((a, b) => b.pricePerNight - a.pricePerNight);
      break;
    case "rating":
      results.sort((a, b) => b.rating - a.rating);
      break;
    case "reviews":
      results.sort((a, b) => b.reviews - a.reviews);
      break;
    default:
      // Best match: rating weighted by log10(reviews) — aligned with Python
      results.sort((a, b) => {
        const scoreA = a.rating * Math.log10(a.reviews + 1);
        const scoreB = b.rating * Math.log10(b.reviews + 1);
        return scoreB - scoreA;
      });
  }

  // Pagination
  const page = Math.max(1, parseInt(pageStr) || 1);
  const limit = Math.min(100, Math.max(1, parseInt(limitStr) || 20));
  const total = results.length;
  const totalPages = Math.ceil(total / limit);
  const start = (page - 1) * limit;
  const paginated = results.slice(start, start + limit);

  res.json({ total, page, limit, totalPages, results: paginated });
});

// GET /api/hotels/featured — top hotels for landing page
router.get("/featured", (req, res) => {
  const featured = [...hotels]
    .sort((a, b) => b.rating * Math.log10(b.reviews + 1) - a.rating * Math.log10(a.reviews + 1))
    .slice(0, 4);
  res.json({ total: featured.length, results: featured });
});

// GET /api/hotels/amenities/all — list all unique amenities
router.get("/amenities/all", (req, res) => {
  const all = [...new Set(hotels.flatMap((h) => h.amenities))].sort();
  res.json(all);
});

// GET /api/hotels/cities/all — list all cities
router.get("/cities/all", (req, res) => {
  const cities = [...new Set(hotels.map((h) => h.city))].sort();
  res.json(cities);
});

// GET /api/hotels/:id/rooms — room types for a hotel
router.get("/:id/rooms", (req, res) => {
  const hotel = hotels.find((h) => h.id === req.params.id);
  if (!hotel) return res.status(404).json({ error: "Hotel not found" });
  res.json({ rooms: hotel.rooms || [] });
});

// GET /api/hotels/:id/reviews — list reviews for a hotel
router.get("/:id/reviews", (req, res) => {
  const hotel = hotels.find((h) => h.id === req.params.id);
  if (!hotel) return res.status(404).json({ error: "Hotel not found" });
  const hotelReviews = allReviews.filter((r) => r.hotelId === req.params.id);
  res.json({ total: hotelReviews.length, reviews: hotelReviews });
});

// POST /api/hotels/:id/reviews — add a review
router.post("/:id/reviews", (req, res) => {
  const hotel = hotels.find((h) => h.id === req.params.id);
  if (!hotel) return res.status(404).json({ error: "Hotel not found" });

  const { author, rating, text } = req.body;
  if (!author || !rating || !text) {
    return res.status(400).json({ error: "Missing required fields: author, rating, text" });
  }

  const numRating = Number(rating);
  if (numRating < 1 || numRating > 5) {
    return res.status(400).json({ error: "Rating must be between 1 and 5" });
  }

  const review = {
    id: "r" + Date.now(),
    hotelId: req.params.id,
    author: author.trim(),
    rating: numRating,
    text: text.trim(),
    date: new Date().toISOString().split("T")[0],
  };

  allReviews.push(review);

  // Update hotel's average rating and review count
  const hotelReviews = allReviews.filter((r) => r.hotelId === req.params.id);
  const avgRating = hotelReviews.reduce((s, r) => s + r.rating, 0) / hotelReviews.length;
  hotel.rating = Math.round(avgRating * 10) / 10;
  hotel.reviews = hotelReviews.length;

  res.status(201).json({ message: "Review added", review });
});

// GET /api/hotels/:id/availability — check date availability
router.get("/:id/availability", (req, res) => {
  const hotel = hotels.find((h) => h.id === req.params.id);
  if (!hotel) return res.status(404).json({ error: "Hotel not found" });

  const { checkIn, checkOut } = req.query;
  if (!checkIn || !checkOut) {
    return res.status(400).json({ error: "checkIn and checkOut are required" });
  }

  const inDate = new Date(checkIn);
  const outDate = new Date(checkOut);
  if (isNaN(inDate) || isNaN(outDate) || outDate <= inDate) {
    return res.status(400).json({ error: "Invalid date range" });
  }

  const overlapping = bookings.filter(
    (b) =>
      b.hotelId === hotel.id &&
      b.status !== "cancelled" &&
      new Date(b.checkIn) < outDate &&
      new Date(b.checkOut) > inDate
  );

  res.json({
    hotelId: hotel.id,
    checkIn,
    checkOut,
    available: overlapping.length === 0,
    overlappingBookings: overlapping.length,
  });
});

// GET /api/hotels/:id — single hotel
router.get("/:id", (req, res) => {
  const hotel = hotels.find((h) => h.id === req.params.id);
  if (!hotel) return res.status(404).json({ error: "Hotel not found" });
  res.json(hotel);
});

module.exports = router;

File: backend/app/routes/hotels.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from sqlalchemy import or_
from datetime import datetime
from typing import Optional
from app.database import get_db
from app.models.hotel import Hotel

router = APIRouter(prefix="/api/hotels", tags=["Hotels"])


def _find_by_public_id(db: Session, public_id: str):
    hotel = db.query(Hotel).filter(Hotel.public_id == public_id).first()
    if not hotel:
        raise HTTPException(status_code=404, detail="Hotel not found")
    return hotel


@router.get("")
def get_hotels(
    city: Optional[str] = None,
    q: Optional[str] = None,
    amenities: Optional[str] = None,
    minPrice: Optional[float] = None,
    maxPrice: Optional[float] = None,
    stars: Optional[int] = None,
    type: Optional[str] = None,
    sort: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    query = db.query(Hotel)

    if city:
        query = query.filter(Hotel.city.ilike(f"%{city}%"))
    if q:
        query = query.filter(
            Hotel.name.ilike(f"%{q}%") | Hotel.description.ilike(f"%{q}%")
        )
    if minPrice is not None:
        query = query.filter(Hotel.price_per_night >= minPrice)
    if maxPrice is not None:
        query = query.filter(Hotel.price_per_night <= maxPrice)
    if stars is not None:
        query = query.filter(Hotel.stars >= stars)
    if type:
        types = [t.strip() for t in type.split(",")]
        query = query.filter(Hotel.hotel_type.in_(types))

    # Amenities filter (Python-side)
    if amenities:
        wanted = [a.strip().lower() for a in amenities.split(",")]
        all_h = query.all()
        filtered = []
        for h in all_h:
            h_amenities = [a.lower() for a in h.amenities]
            if any(a in h_amenities for a in wanted):
                filtered.append(h)
        hotel_ids = [h.id for h in filtered]
        query = db.query(Hotel).filter(Hotel.id.in_(hotel_ids))

    # Sorting
    if sort == "price_asc":
        query = query.order_by(Hotel.price_per_night.asc())
    elif sort == "price_desc":
        query = query.order_by(Hotel.price_per_night.desc())
    elif sort == "rating":
        query = query.order_by(Hotel.rating.desc())
    elif sort == "reviews":
        query = query.order_by(Hotel.reviews_count.desc())

    total = query.count()
    total_pages = (total + limit - 1) // limit
    hotels = query.offset((page - 1) * limit).limit(limit).all()

    return {
        "total": total,
        "page": page,
        "limit": limit,
        "totalPages": total_pages,
        "results": [h.to_dict() for h in hotels],
    }


@router.get("/featured")
def get_featured(db: Session = Depends(get_db)):
    hotels = db.query(Hotel).order_by(Hotel.rating.desc()).limit(6).all()
    return [h.to_dict() for h in hotels]


@router.get("/amenities/all")
def get_all_amenities(db: Session = Depends(get_db)):
    hotels = db.query(Hotel).all()
    all_set = set()
    for h in hotels:
        all_set.update(h.amenities)
    return sorted(all_set)


@router.get("/cities/all")
def get_all_cities(db: Session = Depends(get_db)):
    results = db.query(Hotel.city).distinct().all()
    return sorted([r[0] for r in results if r[0]])


@router.get("/{public_id}")
def get_hotel(public_id: str, db: Session = Depends(get_db)):
    hotel = _find_by_public_id(db, public_id)
    return hotel.to_dict()


@router.get("/{public_id}/rooms")
def get_hotel_rooms(public_id: str, db: Session = Depends(get_db)):
    hotel = _find_by_public_id(db, public_id)
    return {"rooms": hotel.rooms}


@router.get("/{public_id}/reviews")
def get_hotel_reviews(public_id: str):
    return {"reviews": []}


@router.post("/{public_id}/reviews")
def add_hotel_review(public_id: str, body: dict, db: Session = Depends(get_db)):
    _find_by_public_id(db, public_id)
    return {
        "review": {
            "id": f"r{datetime.utcnow().strftime('%y%m%d%H%M%S')}",
            "hotelId": public_id,
            "author": body.get("author", "Anonymous"),
            "rating": body.get("rating", 5),
            "text": body.get("text", ""),
            "date": datetime.utcnow().strftime("%Y-%m-%d"),
        }
    }


@router.get("/{public_id}/availability")
def check_availability(public_id: str, db: Session = Depends(get_db)):
    hotel = _find_by_public_id(db, public_id)
    return {"available": True, "rooms": hotel.rooms}
